utils/workspace.js

```diff
@@ -0,0 +1,205 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+//@ts-check
+
+/**
+ * Use the following command to typescheck this file:
+ * npx tsc --target es2020  --watch --checkjs --noemit --moduleResolution node workspace.js
+ */
+const fs = require('fs');
+const path = require('path');
+const util = require('util');
+const url = require('url');
+
+const readJSON = async (filePath) => JSON.parse(await fs.promises.readFile(filePath, 'utf8'));
+const writeJSON = async (filePath, json) => {
+  await fs.promises.writeFile(filePath, JSON.stringify(json, null, 2) + '\n');
+}
+
+class PWPackage {
+  constructor(descriptor) {
+    this.name = descriptor.name;
+    this.path = descriptor.path;
+    this.files = descriptor.files;
+    this.packageJSONPath = path.join(this.path, 'package.json');
+    this.packageJSON = JSON.parse(fs.readFileSync(this.packageJSONPath, 'utf8'));
+    this.isPrivate = !!this.packageJSON.private;
+  }
+}
+
+class Workspace {
+  /**
+   * @param {string} rootDir
+   * @param {PWPackage[]} packages
+   */
+  constructor(rootDir, packages) {
+    this._rootDir = rootDir;
+    this._packages = packages;
+  }
+
+  packages() { return this._packages; }
+
+  /**
+   * @param {string} version
+   */
+  async setVersion(version) {
+    // 1. update workspace's package.json (playwright-internal) with the new version
+    const workspacePackageJSON = await readJSON(path.join(this._rootDir, 'package.json'));
+    workspacePackageJSON.version = version;
+    await writeJSON(path.join(this._rootDir, 'package.json'), workspacePackageJSON);
+    // 2. make workspace consistent.
+    await this.ensureConsistent();
+  }
+
+  async ensureConsistent() {
+    let hasChanges = false;
+
+    const maybeWriteJSON = async (jsonPath, json) => {
+      const oldJson = await readJSON(jsonPath);
+      if (JSON.stringify(json) === JSON.stringify(oldJson))
+        return;
+      hasChanges = true;
+      console.warn('Updated', jsonPath);
+      await writeJSON(jsonPath, json);
+    };
+
+    const workspacePackageJSON = await readJSON(path.join(this._rootDir, 'package.json'));
+    const packageLockPath = path.join(this._rootDir, 'package-lock.json');
+    const packageLock = JSON.parse(await fs.promises.readFile(packageLockPath, 'utf8'));
+    const version = workspacePackageJSON.version;
+
+    // Make sure package-lock version is consistent with root package.json version.
+    packageLock.version = version;
+    packageLock.packages[""].version = version;
+
+    for (const pkg of this._packages) {
+      // 1. Copy package files.
+      for (const file of pkg.files) {
+        const fromPath = path.join(this._rootDir, file);
+        const toPath = path.join(pkg.path, file);
+        await fs.promises.mkdir(path.dirname(pkg.path), { recursive: true });
+        await fs.promises.copyFile(fromPath, toPath);
```

Comment:
Shouldn't this update `hasChanges`?

Reply:
no, these are part of `.gitignore` and are never committed. This kinda makes sense; otherwise, changing `README.md` would require running this script every time.